CRM/Mosaico/BAO/MosaicoTemplate.php

```diff
@@ -38,20 +38,33 @@ public static function findBaseTemplates() {
 
       $templatesUrl = CRM_Mosaico_Utils::getTemplatesUrl('absolute');
 
+      $templatesLocation[] = array('dir' => $templatesDir, 'url' => $templatesUrl);
+
+      $customTemplatesDir = CRM_Core_BAO_Setting::getItem('Mosaico Preferences', 'mosaico_custom_templates_dir');
+      $customTemplatesUrl = CRM_Core_BAO_Setting::getItem('Mosaico Preferences', 'mosaico_custom_templates_url');
+      if (!is_null($customTemplatesDir) && !is_null($customTemplatesUrl)) {
+          if (is_dir($customTemplatesDir)) {
+            $templatesLocation[] = array('dir' => $customTemplatesDir, 'url' => $customTemplatesUrl);
```

Comment:
If you have a site that goes through migrations (e.g. dev=>prod; prod=>dev; prod to a different domain), then storing an absolute path or absolute URL makes the migration more difficult.

We've started storing paths which are relative to some base, such as `[civicrm.root]/foo` or `[civicrm.files]/bar` or `[cms.root]/foobar`. The `[...]` are evaluated at runtime. This improves portability, and it makes it easier to set default values. (A default like `/var/www/files/mosaico_tpl` is not portable, `[civicrm.files]/mosaico_tpl` would be portable.)

The smallest change would be to pass the settings through `Civi::paths()`, e.g.

```php
$templatesLocation[] = array(
  'dir' => \Civi::paths()->getPath($customTemplatesDir), 
  'url' => \Civi::paths()->getUrl($customTemplatesUrl),
,); 
```

An absolute path or URL will get passed-through, and relative path or URL will have the appropriate substitution.